pkg/session/serve.go

```diff
@@ -246,38 +247,58 @@ func (s *Session) getEvents(ctx context.Context, payload Request) (v1.LeptonEven
 	if !payload.EndTime.IsZero() {
 		endTime = payload.EndTime
 	}
-	var events v1.LeptonEvents
+	var (
+		wg        sync.WaitGroup
+		eventsBuf = make(chan v1.LeptonComponentEvents, len(allComponents))
+	)
+	localCtx, done := context.WithTimeout(ctx, time.Minute)
+	defer done()
 	for _, componentName := range allComponents {
-		currEvent := v1.LeptonComponentEvents{
-			Component: componentName,
-			StartTime: startTime,
-			EndTime:   endTime,
-		}
 		component, err := components.GetComponent(componentName)
 		if err != nil {
 			log.Logger.Errorw("failed to get component",
 				"operation", "GetEvents",
 				"component", componentName,
 				"error", err,
 			)
-			events = append(events, currEvent)
+			eventsBuf <- v1.LeptonComponentEvents{
+				Component: componentName,
+				StartTime: startTime,
+				EndTime:   endTime,
+			}
 			continue
 		}
-		event, err := component.Events(ctx, startTime)
-		if err != nil {
-			if errors.Is(err, query.ErrNoData) {
-				log.Logger.Debugw("no event found", "component", componentName)
-				continue
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			currEvent := v1.LeptonComponentEvents{
+				Component: componentName,
+				StartTime: startTime,
+				EndTime:   endTime,
 			}
+			log.Logger.Debugw("getting events", "component", componentName)
+			event, err := component.Events(localCtx, startTime)
+			if err != nil {
+				if errors.Is(err, query.ErrNoData) {
+					log.Logger.Debugw("no event found", "component", componentName)
+					return
+				}
 
-			log.Logger.Errorw("failed to invoke component events",
-				"operation", "GetEvents",
-				"component", componentName,
-				"error", err,
-			)
-		} else {
-			currEvent.Events = event
-		}
+				log.Logger.Errorw("failed to invoke component events",
+					"operation", "GetEvents",
+					"component", componentName,
+					"error", err,
+				)
+			} else {
+				log.Logger.Debugw("successfully got events", "component", componentName)
+				currEvent.Events = event
+			}
+			eventsBuf <- currEvent
+		}()
+	}
+	wg.Wait()
```

Comment:
Actually we don't need to wait for all goroutines here. We should ensure all gorooutines send result to the channel even if it failed.

Reply:
You have already done this..so we can remove wg here..